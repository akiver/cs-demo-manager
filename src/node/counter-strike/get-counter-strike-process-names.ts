import { isMac } from 'csdm/node/os/is-mac';
import { isWindows } from 'csdm/node/os/is-windows';

export function getCounterStrikeProcessNames() {
  switch (true) {
    case isWindows:
      return ['csgo.exe', 'cs2.exe'];
    case isMac:
      return ['csgo_osx64', 'cs2'];
    default:
      return ['csgo_linux64', 'cs2'];
  }
}
