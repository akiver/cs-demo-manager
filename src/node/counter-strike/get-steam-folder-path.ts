import path from 'node:path';
import fs from 'fs-extra';
import { homedir } from 'node:os';
import { isMac } from 'csdm/node/os/is-mac';
import { isWindows } from 'csdm/node/os/is-windows';
import { getRegistryStringKey } from 'csdm/node/os/windows-registry';

/**
 * Transforms Steam paths from Windows registry.
 * c:/program files (x86)/steam/steam.exe => C:\\program files (x86)\\steam\\steam.exe
 */
function sanitizePathFromRegistry(path: string) {
  return `${path.charAt(0).toUpperCase()}${path.slice(1).replace(/\//gi, '\\')}`;
}

export async function getSteamFolderPath() {
  if (isWindows) {
    const data = await getRegistryStringKey({
      path: 'Software\\Valve\\Steam',
      name: 'SteamPath',
    });
    if (!data) {
      return undefined;
    }

    const steamFolderPath = sanitizePathFromRegistry(data);
    return steamFolderPath;
  }

  if (isMac) {
    return path.join(homedir(), 'Library', 'Application Support', 'Steam');
  }

  const steamHomePath = path.join(homedir(), '.steam', 'steam');
  const isSteamInHomePath = await fs.pathExists(steamHomePath);
  if (isSteamInHomePath) {
    return steamHomePath;
  }

  return path.join(homedir(), '.local', 'share', 'steam');
}
