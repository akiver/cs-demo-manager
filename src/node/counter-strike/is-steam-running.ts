import { isMac } from 'csdm/node/os/is-mac';
import { isWindows } from 'csdm/node/os/is-windows';
import { areProcessesRunning } from 'csdm/node/os/are-processes-running';

export async function isSteamRunning(): Promise<boolean> {
  let processName: string;
  switch (true) {
    case isWindows:
      processName = 'steam.exe';
      break;
    case isMac:
      processName = 'steam_osx';
      break;
    default:
      processName = 'steam';
  }

  const processIsRunning = await areProcessesRunning([processName]);

  return processIsRunning;
}
