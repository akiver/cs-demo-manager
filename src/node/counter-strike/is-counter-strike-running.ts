import { areProcessesRunning } from 'csdm/node/os/are-processes-running';
import { getCounterStrikeProcessNames } from './get-counter-strike-process-names';

export async function isCounterStrikeRunning() {
  const processNames = getCounterStrikeProcessNames();
  const atLeastOneProcessIsRunning = await areProcessesRunning(processNames);

  return atLeastOneProcessIsRunning;
}
