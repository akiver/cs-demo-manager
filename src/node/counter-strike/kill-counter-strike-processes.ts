import { killProcessesByNames } from 'csdm/node/os/kill-processes-by-names';
import { getCounterStrikeProcessNames } from './get-counter-strike-process-names';

export async function killCounterStrikeProcesses() {
  const processNames = getCounterStrikeProcessNames();
  const atLeastOneProcessHasBeenKilled = await killProcessesByNames(processNames);

  return atLeastOneProcessHasBeenKilled;
}
