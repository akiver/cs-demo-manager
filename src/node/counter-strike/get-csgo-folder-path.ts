import fs from 'fs-extra';
import path from 'node:path';
import { getSteamFolderPath } from './get-steam-folder-path';
import { CounterStrikeExecutableNotFound } from './launcher/errors/counter-strike-executable-not-found';
import type { Game } from 'csdm/common/types/counter-strike';

/**
 * Return the path to the "Counter-Strike Global Offensive" folder.
 * https://developer.valvesoftware.com/wiki/Counter-Strike:_Global_Offensive_Game_State_Integration#Locating_CS:GO_Install_Directory
 *
 */
export async function getCsgoFolderPath() {
  const steamFolderPath = await getSteamFolderPath();
  if (steamFolderPath === undefined) {
    return;
  }

  const vdfPath = path.join(steamFolderPath, 'steamapps', 'libraryfolders.vdf');
  const vdfFileExists = await fs.pathExists(vdfPath);
  if (!vdfFileExists) {
    return;
  }

  // Linux / MacOS:
  // "1"		"/home/akiver/Desktop/games"
  // Windows:
  // "1"		"E:\\Steam\\Games"
  const pathRegex = /"(([a-z]:|\/)(.*))"/gi;
  const vdfContent = await fs.readFile(vdfPath);
  const lines = vdfContent.toString().split('\n');
  const libraries = [steamFolderPath];
  for (const line of lines) {
    const matches = pathRegex.exec(line);
    if (matches?.[1] !== undefined) {
      libraries.push(matches[1]);
    }
  }

  for (const library of libraries) {
    const csgoFolderPath = path.join(library, 'steamapps', 'common', 'Counter-Strike Global Offensive');
    const csgoFolderExists = await fs.pathExists(csgoFolderPath);
    if (csgoFolderExists) {
      return csgoFolderPath;
    }
  }
}

export async function getCsgoFolderPathOrThrow(game: Game) {
  const csgoFolderPath = await getCsgoFolderPath();
  if (csgoFolderPath === undefined) {
    throw new CounterStrikeExecutableNotFound(game);
  }

  return csgoFolderPath;
}
