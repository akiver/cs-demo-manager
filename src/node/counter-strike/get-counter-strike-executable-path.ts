import path from 'node:path';
import glob from 'tiny-glob';
import fs from 'fs-extra';
import { Game } from 'csdm/common/types/counter-strike';
import { isMac } from 'csdm/node/os/is-mac';
import { isWindows } from 'csdm/node/os/is-windows';
import { getCsgoFolderPath } from './get-csgo-folder-path';
import { getSteamFolderPath } from './get-steam-folder-path';
import { CounterStrikeExecutableNotFound } from './launcher/errors/counter-strike-executable-not-found';

// Tip: to understand how Steam starts CS you can use the Steam client launch options: -dev -console
// You would be able to see the command used to start CS in the console.
export async function getCounterStrikeExecutablePath(game: Game): Promise<string> {
  const csgoFolderPath = await getCsgoFolderPath();
  if (csgoFolderPath === undefined) {
    throw new CounterStrikeExecutableNotFound(game);
  }

  if (isWindows) {
    let executablePath: string;
    if (game === Game.CS2) {
      executablePath = path.join(csgoFolderPath, 'game', 'bin', 'win64', 'cs2.exe');
    } else {
      executablePath = path.join(csgoFolderPath, 'csgo.exe');
    }

    const executableExists = await fs.pathExists(executablePath);
    if (!executableExists) {
      throw new CounterStrikeExecutableNotFound(game);
    }

    return executablePath;
  }

  if (isMac) {
    const runCsgoScriptPath = path.join(csgoFolderPath, 'csgo.sh');
    const scriptExists = await fs.pathExists(runCsgoScriptPath);
    if (!scriptExists) {
      throw new CounterStrikeExecutableNotFound(game);
    }
    // ! Double quotes are important on macOS
    return `"${runCsgoScriptPath}"`;
  }

  const steamFolderPath = await getSteamFolderPath();
  /**
   * On Linux CS must be launched through a Steam Linux Runtime script.
   * This script takes the path to the game's run script as parameter (csgo.sh for CSGO and cs2.sh for CS2).
   * Trying to run CS directly from the game script will not work because it doesn't export runtime libraries paths
   * that depend on the OS arch.
   */
  const steamScriptName = game === Game.CS2 ? 'SteamLinuxRuntime_sniper/_v2-entry-point' : 'steam-runtime/run.sh';
  const [runSteamScriptPath] = await glob(`**/${steamScriptName}`, {
    cwd: steamFolderPath,
    absolute: true,
    filesOnly: true,
  });
  if (!runSteamScriptPath) {
    throw new CounterStrikeExecutableNotFound(game);
  }
  let gameScriptPath: string;
  if (game === Game.CSGO) {
    gameScriptPath = path.join(csgoFolderPath, 'csgo.sh');
  } else {
    gameScriptPath = path.join(csgoFolderPath, 'game', 'cs2.sh');
  }

  const scriptExists = await fs.pathExists(gameScriptPath);
  if (!scriptExists) {
    throw new CounterStrikeExecutableNotFound(game);
  }

  const command =
    game === Game.CS2
      ? `"${runSteamScriptPath}" --verb=waitforexitandrun -- "${gameScriptPath}"`
      : `"${runSteamScriptPath}" "${gameScriptPath}"`;

  return command;
}
