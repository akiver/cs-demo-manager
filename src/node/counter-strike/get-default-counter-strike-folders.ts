import fs from 'fs-extra';
import path from 'node:path';
import { getCsgoFolderPath } from './get-csgo-folder-path';

/**
 * Detects and returns the "csgo" and "replays" folder where demos are stored.
 */
async function getDefaultFolders(subFolderPaths: string[]) {
  const rootPath = await getCsgoFolderPath();
  if (rootPath === undefined) {
    return [];
  }

  const gameFolderPath = path.join(rootPath, ...subFolderPaths);
  const gameFolderExists = await fs.pathExists(gameFolderPath);
  if (!gameFolderExists) {
    return [];
  }

  const replaysFolderPath = path.join(gameFolderPath, 'replays');
  await fs.mkdirp(replaysFolderPath);

  return [gameFolderPath, replaysFolderPath];
}

export async function getDefaultCsgoFolders() {
  return await getDefaultFolders(['csgo']);
}

export async function getDefaultCs2Folders() {
  return await getDefaultFolders(['game', 'csgo']);
}
