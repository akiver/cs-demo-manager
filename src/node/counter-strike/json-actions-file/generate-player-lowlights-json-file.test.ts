import { fileURLToPath } from 'node:url';
import fs from 'node:fs/promises';
import { describe, expect, it } from 'vitest';
import { Perspective } from 'csdm/common/types/perspective';
import type { PlaybackMatch } from '../../database/watch/get-match-playback';
import { generatePlayerLowlightsJsonFile } from './generate-player-lowlights-json-file';
import { Game } from 'csdm/common/types/counter-strike';

describe(`Generate player's lowlights JSON file`, () => {
  it(`CSGO`, async () => {
    const demoPath = fileURLToPath(new URL('output/csgo-player-lowlights.dem', import.meta.url));
    const fakePlayerSteamId = 'fakePlayerSteamId';
    const fakeOpponentSteamId = 'fakeOpponentSteamId';
    const data: PlaybackMatch = {
      checksum: 'checksum',
      demoPath: demoPath,
      tickrate: 64,
      tickCount: 32000,
      actions: [
        {
          roundNumber: 1,
          tick: 1000,
          opponentSteamId: fakeOpponentSteamId,
          playerSteamId: fakePlayerSteamId,
          playerSlot: null,
          opponentSlot: null,
        },
        {
          roundNumber: 1,
          tick: 1400,
          opponentSteamId: fakeOpponentSteamId,
          playerSteamId: fakePlayerSteamId,
          playerSlot: null,
          opponentSlot: null,
        },
        {
          roundNumber: 1,
          tick: 2000,
          opponentSteamId: fakeOpponentSteamId,
          playerSteamId: fakePlayerSteamId,
          playerSlot: null,
          opponentSlot: null,
        },
        {
          roundNumber: 2,
          tick: 5000,
          opponentSteamId: fakeOpponentSteamId,
          playerSteamId: fakePlayerSteamId,
          playerSlot: null,
          opponentSlot: null,
        },
      ],
    };

    await generatePlayerLowlightsJsonFile({
      game: Game.CSGO,
      actions: data.actions,
      demoPath: data.demoPath,
      tickCount: data.tickCount,
      tickrate: data.tickrate,
      perspective: Perspective.Player,
      beforeDelaySeconds: 2,
      nextDelaySeconds: 4,
      playerVoicesEnabled: true,
    });

    const content = await fs.readFile(`${demoPath}.json`, 'utf8');
    expect(content).toMatchSnapshot();
  });

  it(`CS2`, async () => {
    const demoPath = fileURLToPath(new URL('output/cs2-player-lowlights.dem', import.meta.url));
    const fakePlayerSlot = 4;
    const fakeOpponentPlayerSlot = 8;
    const data: PlaybackMatch = {
      checksum: 'checksum',
      demoPath: demoPath,
      tickrate: 64,
      tickCount: 32000,
      actions: [
        {
          roundNumber: 1,
          tick: 1000,
          opponentSteamId: null,
          playerSteamId: null,
          playerSlot: fakePlayerSlot,
          opponentSlot: fakeOpponentPlayerSlot,
        },
        {
          roundNumber: 1,
          tick: 1400,
          opponentSteamId: null,
          playerSteamId: null,
          playerSlot: fakePlayerSlot,
          opponentSlot: fakeOpponentPlayerSlot,
        },
        {
          roundNumber: 1,
          tick: 2000,
          opponentSteamId: null,
          playerSteamId: null,
          playerSlot: fakePlayerSlot,
          opponentSlot: fakeOpponentPlayerSlot,
        },
        {
          roundNumber: 2,
          tick: 5000,
          opponentSteamId: null,
          playerSteamId: null,
          playerSlot: fakePlayerSlot,
          opponentSlot: fakeOpponentPlayerSlot,
        },
      ],
    };

    await generatePlayerLowlightsJsonFile({
      game: Game.CS2,
      actions: data.actions,
      demoPath: data.demoPath,
      tickCount: data.tickCount,
      tickrate: data.tickrate,
      perspective: Perspective.Player,
      beforeDelaySeconds: 2,
      nextDelaySeconds: 4,
      playerVoicesEnabled: true,
    });

    const content = await fs.readFile(`${demoPath}.json`, 'utf8');
    expect(content).toMatchSnapshot();
  });
});
