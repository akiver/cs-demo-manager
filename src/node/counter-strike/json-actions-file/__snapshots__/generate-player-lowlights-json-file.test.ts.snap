// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generate player's lowlights JSON file > CS2 1`] = `
"[
  {
    "cmd": "tv_listen_voice_indices -1",
    "tick": 64
  },
  {
    "cmd": "tv_listen_voice_indices_h -1",
    "tick": 64
  },
  {
    "cmd": "spec_player 4",
    "tick": 1200
  },
  {
    "cmd": "spec_mode 1",
    "tick": 1200
  },
  {
    "cmd": "spec_player 4",
    "tick": 1700
  },
  {
    "cmd": "spec_mode 1",
    "tick": 1700
  },
  {
    "cmd": "demo_gototick 4872",
    "tick": 2256
  },
  {
    "cmd": "spec_player 4",
    "tick": 4872
  },
  {
    "cmd": "spec_mode 1",
    "tick": 4872
  },
  {
    "cmd": "disconnect",
    "tick": 5256
  }
]"
`;

exports[`Generate player's lowlights JSON file > CSGO 1`] = `
"[
  {
    "cmd": "voice_enable 1",
    "tick": 64
  },
  {
    "cmd": "spec_lock_to_accountid fakeOpponentSteamId",
    "tick": 1200
  },
  {
    "cmd": "spec_player_by_accountid fakeOpponentSteamId",
    "tick": 1200
  },
  {
    "cmd": "spec_lock_to_accountid fakeOpponentSteamId",
    "tick": 1700
  },
  {
    "cmd": "spec_player_by_accountid fakeOpponentSteamId",
    "tick": 1700
  },
  {
    "cmd": "demo_gototick 4872",
    "tick": 2256
  },
  {
    "cmd": "spec_lock_to_accountid fakeOpponentSteamId",
    "tick": 4872
  },
  {
    "cmd": "spec_player_by_accountid fakeOpponentSteamId",
    "tick": 4872
  },
  {
    "cmd": "disconnect",
    "tick": 5256
  }
]"
`;
